src/security/library.cpp

```diff
@@ -0,0 +1,585 @@
+        },
+    .obfuscator =
+        {
+            .key_regex = kDefaultKeyRegex.data(),
+            .value_regex = nullptr,
+        },
+    .free_fn = nullptr,
+};
+
+auto parse_rule_json(std::string_view json) -> dnsec::ddwaf_owned_map {
+  rapidjson::Document document;
+  rapidjson::ParseResult const result =
+      document.Parse(json.data(), json.size());
+  if (!result) {
+    std::string_view error_msg{rapidjson::GetParseError_En(result.Code())};
+    throw std::invalid_argument{"malformed json: " + std::string{error_msg}};
+  }
+  if (!document.IsObject()) {
+    throw std::invalid_argument("invalid json rule (not a json object)");
+  }
+
+  return dnsec::ddwaf_owned_map{
+      dnsec::json_to_object(document, dnsec::kConfigMaxDepth)};
+}
+
+auto read_rule_file(std::string_view filename) -> dnsec::ddwaf_owned_map {
+  std::ifstream rule_file(filename.data(), std::ios::in);
+  if (!rule_file) {
+    throw std::system_error(errno, std::generic_category());
+  }
+
+  // Create a buffer equal to the file size
+  rule_file.seekg(0, std::ios::end);
+  std::string buffer(rule_file.tellg(), '\0');
+  buffer.resize(rule_file.tellg());
+  rule_file.seekg(0, std::ios::beg);
+
+  auto buffer_size = buffer.size();
+  if (buffer_size > static_cast<decltype(buffer_size)>(
+                        std::numeric_limits<std::streamsize>::max())) {
+    throw std::runtime_error{"rule file is too large"};
+  }
+
+  rule_file.read(buffer.data(), static_cast<std::streamsize>(buffer.size()));
+  buffer.resize(rule_file.gcount());
+  rule_file.close();
+
+  return parse_rule_json(buffer);
+}
+
+dnsec::ddwaf_owned_map read_ruleset(
+    std::optional<std::string_view> ruleset_file) {
+  dnsec::ddwaf_owned_map ruleset;
+  if (ruleset_file) {
+    try {
+      ruleset = read_rule_file(*ruleset_file);
+    } catch (const std::exception &e) {
+      throw std::runtime_error(std::string{"failed to read ruleset "} + "at " +
+                               std::string{*ruleset_file} + ": " + e.what());
+    }
+  } else {
+    try {
+      ruleset = parse_rule_json(
+          std::string_view{gRecommendedJsonData, gRecommendedJsonSize});
+    } catch (const std::exception &e) {
+      throw std::runtime_error{
+          "failed to parse embedded recommended ruleset: " +
+          std::string{e.what()}};
+    }
+  }
+
+  return ruleset;
+}
+
+int ddwaf_log_level_to_nginx(DDWAF_LOG_LEVEL level) noexcept {
+  switch (level) {
+    case DDWAF_LOG_TRACE:
+    case DDWAF_LOG_DEBUG:
+      return NGX_LOG_DEBUG;
+    case DDWAF_LOG_INFO:
+      return NGX_LOG_INFO;
+    case DDWAF_LOG_WARN:
+      return NGX_LOG_WARN;
+    case DDWAF_LOG_ERROR:
+      return NGX_LOG_ERR;
+    default:
+      return NGX_LOG_NOTICE;
+  }
+}
+
+DDWAF_LOG_LEVEL ngx_log_level_to_ddwaf(int level) noexcept {
+  switch (level) {
+    case NGX_LOG_DEBUG:
+      return DDWAF_LOG_DEBUG;
+    case NGX_LOG_INFO:
+    case NGX_LOG_NOTICE:
+      return DDWAF_LOG_INFO;
```

Comment:
Some libraries are setting WAF INFO logs to DEBUG as they can be noisy and provide little value to the customer.

Reply:
see my comment before about debug logs